solidity/contracts/modules/MultisigModule.sol

```diff
@@ -0,0 +1,349 @@
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Updates the commitment to the validator set for `_domain`.
+     * @param _domain The remote domain of the validator set.
+     * @return The commitment to the validator set for `_domain`.
+     */
+    function _updateCommitment(uint32 _domain) internal returns (bytes32) {
+        address[] memory _validators = validators(_domain);
+        uint256 _threshold = threshold[_domain];
+        bytes32 _commitment = keccak256(
+            abi.encodePacked(_threshold, _validators)
+        );
+        commitment[_domain] = _commitment;
+        return _commitment;
+    }
+
+    /**
+     * @notice Verifies the merkle proof of `_message` against the provided
+     * checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function _verifyMerkleProof(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) internal pure returns (bool) {
+        // calculate the expected root based on the proof
+        bytes32 _calculatedRoot = MerkleLib.branchRoot(
+            _message.id(),
+            _metadata.proof(),
+            // TODO: The leaf index may not be the same as the nonce if we choose to go
+            // with modular storage for outbound messages.
+            _message.nonce()
+        );
+        return _calculatedRoot == _metadata.root();
+    }
+
+    /**
+     * @notice Verifies that a quorum of the origin domain's validators signed
+     * the provided checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function _verifyValidatorSignatures(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) internal view returns (bool) {
+        uint256 _threshold = _metadata.threshold();
+        bytes32 _digest;
+        {
+            uint32 _origin = _message.origin();
+            bytes32 _commitment = keccak256(
+                abi.encodePacked(_threshold, _metadata.validators())
+            );
+            // Ensures _validators is sorted by ascending address.
+            require(_commitment == commitment[_origin], "!commitment");
+            _digest = _getCheckpointDigest(_metadata, _origin);
+        }
+        uint256 _validatorIndex = 0;
+        // looking for signers within validators
+        // assuming that both validators and signatures are sorted
+        for (uint256 i = 0; i < _threshold; ++i) {
+            address _signer = ECDSA.recover(_digest, _metadata.signatureAt(i));
+            // looping through remaining validators to find a match
+            for (
+                ;
+                _validatorIndex < _threshold &&
+                    _signer != _metadata.validatorAt(_validatorIndex);
+                ++_validatorIndex
+            ) {}
```

Comment:
I did not, that's an interesting idea. We could definitely do that, but I would expect threshold to always be pretty large wrt validatorCount, at least 50%, and so my intuition is that it would be less gas efficient than iterating